ElectronClient/app.js

```diff
@@ -302,6 +313,17 @@ class Application extends BaseApplication {
 			this.updateMenuItemStates(newState);
 		}
 
+		if (['THEME_TOGGLE'].indexOf(action.type) >= 0) {
+			Setting.setValue('isNightMode', newState.isNightMode);
```

Comment:
Do you need to have `isNightMode` as part of the state?

Isn't the theme already stored as a setting and so could we instead just compare the current theme with the preferred light theme to deduce this?

Reply:
I think here I have to explain the functionality of `Night Mode` button in detail as this is similar question asked by Laurent also.

The Night Mode button has it's own state and setting variable `isNightMode` when we click on the button the state of button will be changed to the opposite value of it's previous state and also the updated state value will be set to `isNightMode` setting variable in `Setting.js` after this changes we want to change theme so based on the state of Night Mode Button we change the theme.

If `isNightMode` state is `true` the theme should set to the value of `preferredDarkTheme` from `Setting.js` and similarly if `isNightMode` state is `false` the theme should set to the value of `preferredLightTheme` from `Setting.js`

Along with this we also need to change the Button icon so we can know what current state is for this purpose also we need the state variable so based on the `isNightMode` state the icon will toggle between Moon and Sun icon.

And when we close the app and restart again the button should be in the last active state so for that we will set `isNightMode` state to the value of `isNightMode` variable from `Setting.js` using `THEME_TOGGLE_SET` reducer.

Again I'm saying here this button is completely similar to `Toggle sidebar` and `Toggle notelist` buttons these button toggle visibility of panes and `Night Mode` button toggles between `preferredDarkTheme` and `preferredLightTheme` so I simply followed the same implementation as of  `Toggle notelist` button where state is also used.

I hope my explanation will answer to all your questions why I've used the state for `Night Mode` button

Reply:
Thanks for the explanation.  However I am still not convinced.  I also don't like the behaviour that my `theme` setting is clobbered when I press the night/day button.

I think the feature can probably be simplified:

- In the settings screen:
    - hide the `theme` setting from the user
    - permanently show the `auto-detect`, `preferred light theme` and `preferred dark theme` settings
- In the application
    - when the user presses `night/day` button, toggle the `theme` between the preferred light and dark settings.

This is much simpler, and doesn't need a new setting.  Perhaps I misunderstand the feature, is there a flaw in my suggestion?

Also, when `auto-detect` is enabled, should we disable or hide the `night/day` button?

The comparison to `Toggle sidebar` and `Toggle notelist` is correct in that they all toggle, but I don't think it is a good comparison because you are using four settings for night mode (`theme`, `preferred light theme`, `preferred dark theme`, `night mode`)  where the sidebar and notelist use just one.

Reply:
There was already discussion going on here: https://discourse.joplinapp.org/t/night-mode-button/6581 about implementation of Night Mode button. Here is the Laurent's response:
> Preferred dark theme should always be visible
    Preferred light theme should always be visible
    If auto-detect theme is “on” - hide theme dropdown
    If auto-detect theme is “off” - show theme dropdown

So I implemented in the same way!

I was looking at the other apps which uses auto-detect feature and I found GitHub Desktop app here is how they implemented showing the demo in GIF:
![theme](https://user-images.githubusercontent.com/15380913/79061570-2334b800-7caf-11ea-96c7-5fe087517f57.gif)

Here I observed that on enabling auto-detect option theme set to system theme (Dark in my case) and manually changing the theme automatically disabled the auto-detect option.

> This is much simpler, and doesn't need a new setting. Perhaps I misunderstand the feature, is there a flaw in my suggestion?

I agree with you about removing `theme` option, based on your suggestions I have some changes in mind I would like to share:

 - For `auto-detect` enabled
   - Theme should match with the system theme can be DARK and LIGHT only
   - Hide preferred theme options
   - Night Mode button should toggle between DARK and LIGHT theme only

 - For `auto-detect` disabled 
   - Show preferred theme options
   - Night Mode button should toggle between preferred Dark/Light themes

> Also, when auto-detect is enabled, should we disable or hide the night/day button?

I don't think killing one feature over another feature is good idea. When we work in night time we prefer dark theme and during day time we prefer light theme so `Night Mode` button is a nice handy feature that we can easily toggle between light and dark theme. This is just a my opinion I might be wrong here. 

>The comparison to Toggle sidebar and Toggle notelist is correct in that they all toggle, but I don't think it is a good comparison because you are using four settings for night mode (theme, preferred light theme, preferred dark theme, night mode) where the sidebar and notelist use just one.

Yea I agree with you. I was trying to say that the implementation of Night Mode button is similar to the Toggle sidebar and Toggle notelist, the changes are here:
```javascript
if (['NOTELIST_VISIBILITY_TOGGLE', 'NOTELIST_VISIBILITY_SET'].indexOf(action.type) >= 0) {
			Setting.setValue('noteListVisibility', newState.noteListVisibility);
		}
```

```javascript
if (['NIGHT_MODE_TOGGLE'].indexOf(action.type) >= 0) {
			Setting.setValue('isNightMode', newState.isNightMode);
if (newState.isNightMode) {
				const activeDarkTheme = Setting.value('preferredDarkTheme');
				Setting.setValue('theme', activeDarkTheme);
			} else {
				const activeLightTheme = Setting.value('preferredLightTheme');
				Setting.setValue('theme', activeLightTheme);
			}
		}
```
Toggle notelist changes `noteListVisibility` setting and Night Mode button changes `isNightMode` and `theme` setting. So the whole point of comparison is the build test! I don't think we need any specific build tests for the changes I've made.

Thanks for your time and consideration 😃

Reply:
> I was looking at the other apps which uses auto-detect feature and I found GitHub Desktop app here is how they implemented showing the demo in GIF

Good thinking.  The example looks good.

> There was already discussion going on here: https://discourse.joplinapp.org/t/night-mode-button/6581 about implementation

Ahh thank you.  I will continue this discussion there, as any change to spec will need Laurent's approval in any case.

> I don't think killing one feature over another feature is good idea.

Ok, I don't mind either way for this feature.

Reply:
Thanks for your kind words.

I'm trying to get Laurent's attention here but seems like he is busy. I hope my code is clean now at least for the current specs mentioned in the issue: #2665

Reply:
And thanks for your patience.

There is no need to attract Laurent's attention, he is very busy, but he is aware of all PRs and will attend to them in due course.  Also please be assured, that even if the specs for this feature are changed (or not), it would be noted and it would not affect your assessment for GSOC.

Reply:
>Also please be assured, that even if the specs for this feature are changed (or not), it would be noted and it would not affect your assessment for GSOC.

Thank you so much!! That's a big relief for me 😀